File: mobile/src/main/java/com/siliconlabs/bledemo/toolbars/LoggerFragment.java
package com.siliconlabs.bledemo.toolbars;


import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Handler;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;

import com.siliconlabs.bledemo.R;
import com.siliconlabs.bledemo.adapters.LogAdapter;
import com.siliconlabs.bledemo.log.Log;
import com.siliconlabs.bledemo.services.ShareLogServices;
import com.siliconlabs.bledemo.utils.Constants;

import java.util.ArrayList;
import java.util.List;

public class LoggerFragment extends Fragment {

    private LogAdapter adapter;
    private RecyclerView logRV;
    private Handler handler;
    private boolean allowRefreshScrollBottom = true;
    private boolean isFiltering = false;

    public static String IS_FILTERING_EXTRA = "IS_FILTERING_EXTRA";
    public static String FILTERING_PHRASE_EXTRA = "FILTERING_PHRASE_EXTRA";

    private String filteringPhrase;

    public LoggerFragment() {
    }

    private ToolbarCallback toolbarCallback;

    private static final int LOG_UPDATE_PERIOD = 2000;
    private final Runnable logUpdater = new Runnable() {

        @Override
        public void run() {
            adapter.notifyDataSetChanged();
            handler.postDelayed(logUpdater, LOG_UPDATE_PERIOD);
            if (allowRefreshScrollBottom) {
                logRV.scrollToPosition(adapter.getItemCount() - 1);
            }
            android.util.Log.d("Log_Updater","RUN");
        }
    };

    public LoggerFragment setCallback(ToolbarCallback toolbarCallback) {
        this.toolbarCallback = toolbarCallback;
        return this;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_log, container, false);
        ImageView closeBtn = view.findViewById(R.id.imageview_close);
        final Button shareBtn = view.findViewById(R.id.button_share);
        Button clearBtn = view.findViewById(R.id.textview_clear_log);
        Button filterBtn = view.findViewById(R.id.textview_filter_log);
        final LinearLayout searchBoxLL = view.findViewById(R.id.search_log_container);
        EditText searchLogET = view.findViewById(R.id.edit_text_search);
        ImageView clearLogIV = view.findViewById(R.id.search_log_clear);
        final View logSeparator = view.findViewById(R.id.log_separator);
        setSearchBox(searchLogET, clearLogIV);


        closeBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                toolbarCallback.close();
                stopLogUpdater();
            }
        });
        clearBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Constants.LOGS.clear();
                adapter.setLogList(Constants.LOGS);
                adapter.notifyDataSetChanged();
            }
        });
        filterBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                searchBoxLL.setVisibility(searchBoxLL.getVisibility() == View.VISIBLE ? View.GONE : View.VISIBLE);
                logSeparator.setVisibility(logSeparator.getVisibility() == View.VISIBLE ? View.GONE : View.VISIBLE);
            }
        });
        shareBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(getContext(), ShareLogServices.class);
                intent.putExtra(IS_FILTERING_EXTRA,isFiltering);
                if(isFiltering) {
                    intent.putExtra(FILTERING_PHRASE_EXTRA, filteringPhrase);
                }
                getActivity().startService(intent);
            }
        });
        logRV = view.findViewById(R.id.recyclerview_log);
        LinearLayoutManager llm = new LinearLayoutManager(getActivity());
        llm.setStackFromEnd(true);
        logRV.setLayoutManager(llm);
        logRV.setAdapter(adapter);
        adapter.setLogList(Constants.LOGS);
        adapter.notifyDataSetChanged();
        final List<Log> filtered = new ArrayList<>();

        searchLogET.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                filtered.clear();
                for (Log log : Constants.LOGS) {
                    if (log.getLogInfo().toLowerCase().contains(s.toString().toLowerCase())) {
                        filtered.add(log);
                    }
                }

                if(s.toString().length() <= 0) {
                    adapter.setLogList(Constants.LOGS);
                    isFiltering = false;
                } else {
                    adapter.setLogList(filtered);
                    isFiltering = true;
                    filteringPhrase = s.toString();
                }
                adapter.notifyDataSetChanged();

            }
        });

        logRV.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);

                if(!recyclerView.canScrollVertically(1)) allowRefreshScrollBottom = true;
                else allowRefreshScrollBottom = false;

            }
        });

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        runLogUpdater();
    }

    @Override
    public void onPause() {
        super.onPause();
        stopLogUpdater();
        android.util.Log.d("OnPause", "LoggerFragment");
    }

    public void setAdapter(LogAdapter adapter) {
        this.adapter = adapter;
    }

    private void setSearchBox(final EditText searchEditText, final ImageView clearImageView) {

        searchEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (count <= 0) clearImageView.setVisibility(View.GONE);
                else clearImageView.setVisibility(View.VISIBLE);
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        clearImageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                searchEditText.getText().clear();
            }
        });

    }

    public void scrollToEnd() {
        if (logRV != null && adapter.getItemCount() > 1) {
            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    logRV.scrollToPosition(adapter.getItemCount() - 1);
                }
            }, 200);
        }
    }

    public void runLogUpdater() {
        if (handler == null) handler = new Handler();
        handler.removeCallbacks(logUpdater);
        handler.postDelayed(logUpdater, LOG_UPDATE_PERIOD);
        android.util.Log.d("Log_Updater","START");
    }

    public void stopLogUpdater() {
        handler.removeCallbacks(logUpdater);
        android.util.Log.d("Log_Updater","STOP");
    }

    public LogAdapter getAdapter() {
        return this.adapter;
    }
}

File: mobile/src/main/java/com/siliconlabs/bledemo/Browser/Fragments/LoggerFragment.kt
package com.siliconlabs.bledemo.Browser.Fragments

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.siliconlabs.bledemo.Browser.Adapters.LogAdapter
import com.siliconlabs.bledemo.Browser.Models.Logs.Log
import com.siliconlabs.bledemo.Browser.Services.ShareLogServices
import com.siliconlabs.bledemo.Browser.ToolbarCallback
import com.siliconlabs.bledemo.R
import com.siliconlabs.bledemo.Utils.Constants
import kotlinx.android.synthetic.main.fragment_log.*
import java.util.*

class LoggerFragment : Fragment() {
    var adapter: LogAdapter? = null
    private var handler: Handler? = null
    private var allowRefreshScrollBottom = true
    private var isFiltering = false
    private var filteringPhrase: String? = null
    private var toolbarCallback: ToolbarCallback? = null


    private val logUpdater: Runnable = object : Runnable {
        override fun run() {
            adapter?.notifyDataSetChanged()
            handler?.postDelayed(this, LOG_UPDATE_PERIOD.toLong())
            if (allowRefreshScrollBottom) rv_log?.scrollToPosition(adapter?.itemCount!! - 1)
            android.util.Log.d("Log_Updater", "RUN")
        }
    }

    fun setCallback(toolbarCallback: ToolbarCallback?): LoggerFragment {
        this.toolbarCallback = toolbarCallback
        return this
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_log, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setSearchBox(et_search, iv_search_log_clear)
        iv_close.setOnClickListener {
            toolbarCallback?.close()
            stopLogUpdater()
        }

        btn_clear_log.setOnClickListener {
            Constants.LOGS.clear()
            adapter?.setLogList(Constants.LOGS)
            adapter?.notifyDataSetChanged()
        }

        btn_filter_log.setOnClickListener {
            ll_search_log_container.visibility = if (ll_search_log_container.visibility == View.VISIBLE) View.GONE else View.VISIBLE
            log_separator.visibility = if (log_separator.visibility == View.VISIBLE) View.GONE else View.VISIBLE
        }

        btn_share.setOnClickListener {
            val intent = Intent(context, ShareLogServices::class.java).apply {
                putExtra(IS_FILTERING_EXTRA, isFiltering)
                if (isFiltering) putExtra(FILTERING_PHRASE_EXTRA, filteringPhrase)
            }
            activity?.startService(intent)
        }

        val llm = LinearLayoutManager(activity)
        llm.stackFromEnd = true
        rv_log.layoutManager = llm
        rv_log.adapter = adapter
        adapter?.setLogList(Constants.LOGS)
        adapter?.notifyDataSetChanged()

        val filtered: MutableList<Log> = ArrayList()
        et_search.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                filtered.clear()
                for (log in Constants.LOGS) {
                    if (log.logInfo.toLowerCase().contains(s.toString().toLowerCase())) {
                        filtered.add(log)
                    }
                }
                if (s.toString().isEmpty()) {
                    adapter?.setLogList(Constants.LOGS)
                    isFiltering = false
                } else {
                    adapter?.setLogList(filtered)
                    isFiltering = true
                    filteringPhrase = s.toString()
                }
                adapter?.notifyDataSetChanged()
            }
        })

        rv_log.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                allowRefreshScrollBottom = !recyclerView.canScrollVertically(1)
            }
        })

    }

    override fun onResume() {
        super.onResume()
        runLogUpdater()
    }

    override fun onPause() {
        super.onPause()
        stopLogUpdater()
        android.util.Log.d("OnPause", "LoggerFragment")
    }

    private fun setSearchBox(searchEditText: EditText, clearImageView: ImageView) {
        searchEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (count <= 0) clearImageView.visibility = View.GONE else clearImageView.visibility = View.VISIBLE
            }

            override fun afterTextChanged(s: Editable) {}
        })
        clearImageView.setOnClickListener { searchEditText.text.clear() }
    }

    fun scrollToEnd() {
        if (rv_log != null && adapter?.itemCount!! > 1) {
            Handler().postDelayed({ rv_log?.scrollToPosition(adapter?.itemCount!! - 1) }, 200)
        }
    }

    fun runLogUpdater() {
        handler = handler ?: Handler()
        handler?.removeCallbacks(logUpdater)
        handler?.postDelayed(logUpdater, LOG_UPDATE_PERIOD.toLong())
        android.util.Log.d("Log_Updater", "START")
    }

    fun stopLogUpdater() {
        handler?.removeCallbacks(logUpdater)
        android.util.Log.d("Log_Updater", "STOP")
    }

    companion object {
        const val IS_FILTERING_EXTRA = "IS_FILTERING_EXTRA"
        const val FILTERING_PHRASE_EXTRA = "FILTERING_PHRASE_EXTRA"
        private const val LOG_UPDATE_PERIOD = 2000
    }
}